File: Backend_NodeJS/models/reviewModel.js
import { execute } from "../config/db.js";

export default class reviewModel {
    // Hàm lấy danh sách đánh giá của bác sĩ theo Ma_bac_si
    static async getReviewsByDoctorId(ma_bac_si) {
        try {
            // Thực hiện Double JOIN: danh_gia -> benh_nhan -> nguoi_dung
            const query = `
                SELECT 
                    dg.*, 
                    nd.Ten_nguoi_dung AS Ten_benh_nhan, 
                    nd.Anh_dai_dien 
                FROM danh_gia dg
                LEFT JOIN benh_nhan bn ON dg.Ma_benh_nhan = bn.Ma_benh_nhan
                LEFT JOIN nguoi_dung nd ON bn.Ma_nguoi_dung = nd.Ma_nguoi_dung
                WHERE dg.Ma_bac_si = ?
                ORDER BY dg.Ngay_tao DESC
            `;
            const [rows] = await execute(query, [ma_bac_si]);
            return rows;
        } catch (error) {
            throw new Error("Lỗi truy vấn danh sách đánh giá: " + error.message);
        }
    }

    static async createReview(data){
        try{
            console.log(data);
            const [checkDone] = await execute(
                `SELECT Trang_thai_lich_hen
                FROM lich_hen
                WHERE Ma_lich_hen = ?`
                ,[data.Ma_lich_hen]
            );

            if(checkDone[0].Trang_thai_lich_hen != 'done') throw new Error("Lịch hẹn này chưa hoàn thành.");

            
            const [checkReview] = await execute(`
                    SELECT *
                    FROM danh_gia
                    WHERE Ma_lich_hen = ?
                `,[data.Ma_lich_hen]);

            if(checkReview.length > 0) throw new Error("Bạn đã đáng giá lần khám này rồi.");

            const sql = `
                INSERT INTO danh_gia(
                    Ma_bac_si,
                    Ma_lich_hen,
                    Ma_benh_nhan,
                    So_sao,
                    Noi_dung,
                    Ngay_tao
                )
                VALUES(?,?,?,?,?,?)
            `;

            const [result] = await execute(sql,[data.Ma_bac_si, data.Ma_lich_hen, data.Ma_benh_nhan, data.So_sao, data.Noi_dung, new Date()]);

            return result.insertId ? result.insertId : null;
        }catch(error){
            throw new Error("Lỗi thêm đánh giá: " + error.message);
        }
    }
}

File: Fronend_Flutter/ung_dung_dat_lich_kham/lib/Models/review_model.dart
class ReviewModel {
  final int id;
  final int doctorId;
  final String patientName;
  final String? patientAvatar;
  final int rating;
  final String content;
  final String createdAt;

  ReviewModel({
    required this.id,
    required this.doctorId,
    required this.patientName,
    this.patientAvatar,
    required this.rating,
    required this.content,
    required this.createdAt,
  });

  factory ReviewModel.fromJson(Map<String, dynamic> json) {
    return ReviewModel(
      id: json['Ma_danh_gia'],
      doctorId: json['Ma_bac_si'],
      patientName: json['Ten_benh_nhan'] ?? 'Bệnh nhân ẩn danh',
      patientAvatar: json['Anh_dai_dien'],
      rating: json['So_sao'] ?? 5,
      content: json['Noi_dung'] ?? '',
      createdAt: json['Ngay_tao'] ?? '',
    );
  }
}